In [1]:
import pandas as pd
import numpy as np

# Dropping Columns in a DataFrame

In [16]:
df = pd.read_csv("data/BL-Flickr-Images-Book.csv")
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [7]:
to_drop = ['Edition Statement', 
          'Corporate Author', 
          'Former owner', 
          'Engraver', 
          'Contributors', 
          'Issuance type', 
          'Shelfmarks']

In [10]:
df.drop(labels = to_drop, inplace=True, axis = 1)
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author  Corporate Contributors  \
0      A. A.                     NaN   
1  A., A. A.                     NaN   
2  A., A. A.                     NaN   
3  A., E. S.                     NaN   
4  A., E. S.                     NaN   

                                          Flickr URL  
0  http://www.flickr.com/photos/britishlibrary/ta...  
1  http://www.flickr.com/photos/britishlibrary/ta...  
2  http://www.flickr.com/photos/britishlibrary/ta...  
3  http://www.flickr.com/photos/britishlibrary/ta...  
4  http://www.flickr.com/photos/britishlibrary/ta...

In [17]:
# to droped columns directly passed to the columns parameter
df.drop(columns = to_drop, inplace = True)
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author  Corporate Contributors  \
0      A. A.                     NaN   
1  A., A. A.                     NaN   
2  A., A. A.                     NaN   
3  A., E. S.                     NaN   
4  A., E. S.                     NaN   

                                          Flickr URL  
0  http://www.flickr.com/photos/britishlibrary/ta...  
1  http://www.flickr.com/photos/britishlibrary/ta...  
2  http://www.flickr.com/photos/britishlibrary/ta...  
3  http://www.flickr.com/photos/britishlibrary/ta...  
4  http://www.flickr.com/photos/britishlibrary/ta...

> If you known in advance which columns you'd like to retain, another options is to pass them to the `usecols` argument of `pd.read_csv`.

# Changing the Index of a `DataFrame`

A Pandas Index extends the functionality of Numpy Arrays to allow for more versatile slicing and labeling. In many cases, it is helpful to use a uniquely valued identifying field of the data as its index.

In [26]:
df["Identifier"].is_unique
df = df.set_index('Identifier')
# or df.set_index("Identifier", inplace = True)
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

            Corporate Contributors  \
Identifier                           
206                            NaN   
216                            NaN   
218                            NaN   
472                            NaN   
480                            NaN   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

In [27]:
df.index

Int64Index([    206,     216,     218,     472,     480,     481,     519,
                667,     874,    1143,
            ...
            4117751, 4117752, 4156359, 4157746, 4157862, 4158088, 4158128,
            4159563, 4159587, 4160339],
           dtype='int64', name='Identifier', length=8287)

> Unlike primary keys in SQL, a Pandas `Index` doesn't make any guarantee of being unique, although many indexing and merging operations will notice a speedup in runtime if it is.


In [28]:
df.loc[206]

Place of Publication                                                 London
Date of Publication                                             1879 [1878]
Publisher                                                  S. Tinsley & Co.
Title                                     Walter Forbes. [A novel.] By A. A
Author                                                                A. A.
Corporate Contributors                                                  NaN
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

In [31]:
df.loc[206, 'Title']

'Walter Forbes. [A novel.] By A. A'

# Tidying up Fields in the Data



In [34]:
df.columns

Index(['Place of Publication', 'Date of Publication', 'Publisher', 'Title',
       'Author', 'Corporate Contributors', 'Flickr URL'],
      dtype='object')

In [35]:
df.loc[1905:, "Date of Publication"].head(10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

## Tasks
- remove the extra dates in square brackets, wherever present: 1879[1878]  
- convert date ranges to their 'start date', wherever present: 1860-63; 1839,38-54  
- completely remove the dates we are not certain about and replace them with Numpy's NaN: [1897?]  
- convert the string nan to Numpy's NaN value.

In [37]:
regex = r'^(\d{4})'
# the parentheses denote a capturing group

In [38]:
extr = df['Date of Publication'].str.extract(r'^(\d{4})', expand = False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

In [39]:
df['Date of publication'] = pd.to_numeric(extr)
df['Date of publication'].dtype

dtype('float64')

In [41]:
df["Date of Publication"].isnull().sum()/len(df)

0.021841438397490046

# Combining `str` Methods with Numpy to clean Columns